src/main/java/org/opensearch/neuralsearch/search/query/HybridCollectorManager.java

```diff
@@ -461,6 +477,37 @@ private ReduceableSearchResult reduceSearchResults(final List<ReduceableSearchRe
         };
     }
 
+    /**
+     * Get maximum subquery results count to be collected from each shard.
+     * @param searchContext search context that contains pagination depth
+     * @return results size to collected
+     */
+    private static int getSubqueryResultsRetrievalSize(final SearchContext searchContext) {
+        HybridQuery hybridQuery = unwrapHybridQuery(searchContext.query());
+        int paginationDepth = hybridQuery.getQueryContext().getPaginationDepth();
+
+        // Switch to from+size retrieval size during standard hybrid query execution.
+        if (searchContext.from() == 0) {
+            return searchContext.size();
+        }
+        return paginationDepth;
+    }
+
+    /**
+     * Unwraps a HybridQuery from either a direct query or a nested BooleanQuery
+     */
+    private static HybridQuery unwrapHybridQuery(Query query) {
+        HybridQuery hybridQuery;
+        // In case of nested fields and alias filter, hybrid query is wrapped under bool query and lies in the first clause.
+        if (query instanceof BooleanQuery) {
+            BooleanQuery booleanQuery = (BooleanQuery) query;
+            hybridQuery = (HybridQuery) booleanQuery.clauses().get(0).getQuery();
```

Comment:
we need more checks here, I think what you're trying to do is similar to this method in QPS https://github.com/opensearch-project/neural-search/blob/main/src/main/java/org/opensearch/neuralsearch/search/query/HybridQueryPhaseSearcher.java#L87-L103. 
Can you extract some common code to the utility class, this one is a good candidate: https://github.com/opensearch-project/neural-search/blob/main/src/main/java/org/opensearch/neuralsearch/util/HybridQueryUtil.java

Reply:
In the whole hybrid query work flow, the query phase searcher will execute first and will check these conditions thoroughly. In hybrid collector manager where I just want to get the hybrid query object, I think just checking for the instance of boolean query will work. Because I add those checks of QPS here also, then in the whole workflow it will be redundant and unecessorily impact performance. WDYT?

Reply:
How we ended up with the HQ not being the top query (I assume that's the case because of the unwrap method) if the QPS should do such processing and unwrap everything at the first place? Is this some sort of a edge case?

Reply:
Regarding the code organization question: I'm in favor of extracting common code, it should not add lot of overhead if we have 2-3 if or instance of checks, but we can eliminate potential problem of having different logic in different places for HQ.